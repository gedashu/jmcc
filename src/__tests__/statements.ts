import JMCC from '../core';

const jmcc = new JMCC();

describe('события', () => {
  test('обычные', () => {
    const source = `
      event<player_join> {
        player::message("hello")
      }

      event<player_quit> {
        player::message("bye")
      }
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('с break', () => {
    const source = `
      event<player_join> {
        player::message("hello")

        break
      }
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });
});

describe('функции', () => {
  test('обычные', () => {
    const source = `
      function func() {}

      func()
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('с аргументами', () => {
    const source = `
      function func(a, b) {}

      func(1, 2)
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('с опциональными аргументами', () => {
    const source = `
      function func(a, b = 0) {}

      func(1)
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('с именными аргументами', () => {
    const source = `
      function func(a, b) {}

      func(a = 1, b = 2)
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('процессы', () => {
    const source = `
      process func(a, b) {}

      func(1, 2)
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('поздние', () => {
    const source = `
      func(1, 2)

      function func(a, b) {}
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });
});

describe('if-блоки', () => {
  test('обычные', () => {
    const source = `
      var a = 1

      if (variable::in_range(a, 1, 2)) {
        player::message("true")
      }
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('not', () => {
    const source = `
      var a = 1

      if not (variable::in_range(a, 1, 2)) {
        player::message("false")
      }
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('с else-блоком', () => {
    const source = `
      var a = 1

      if (variable::in_range(a, 1, 2)) {
        player::message("true")
      } else {
        player::message("false")
      }
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });

  test('с одним выражением', () => {
    const source = `
      var a = 1

      if (variable::in_range(a, 1, 2))
        player::message("true")
      else
        player::message("false")
    `;

    expect(jmcc.compileToJson(source)).toMatchSnapshot();
  });
});
