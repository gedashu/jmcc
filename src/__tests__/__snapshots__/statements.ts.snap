// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if-блоки not 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "if_variable_in_range",
          "is_inverted": true,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "false",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [
            Object {
              "name": "value",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "min",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "max",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`if-блоки обычные 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "if_variable_in_range",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "true",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [
            Object {
              "name": "value",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "min",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "max",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`if-блоки с else-блоком 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "if_variable_in_range",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "true",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [
            Object {
              "name": "value",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "min",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "max",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "else",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "false",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`if-блоки с одним выражением 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "if_variable_in_range",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "true",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [
            Object {
              "name": "value",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "min",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "max",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "else",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "false",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`события обычные 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "player_join",
      "operations": Array [
        Object {
          "action": "player_send_message",
          "values": Array [
            Object {
              "name": "messages",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "text": "hello",
                    "type": "text",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "merging",
              "value": Object {
                "enum": "SPACES",
                "type": "enum",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
    Object {
      "event": "player_quit",
      "operations": Array [
        Object {
          "action": "player_send_message",
          "values": Array [
            Object {
              "name": "messages",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "text": "bye",
                    "type": "text",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "merging",
              "value": Object {
                "enum": "SPACES",
                "type": "enum",
              },
            },
          ],
        },
      ],
      "position": 1,
      "type": "event",
    },
  ],
}
`;

exports[`события с break 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "player_join",
      "operations": Array [
        Object {
          "action": "player_send_message",
          "values": Array [
            Object {
              "name": "messages",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "text": "hello",
                    "type": "text",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "merging",
              "value": Object {
                "enum": "SPACES",
                "type": "enum",
              },
            },
          ],
        },
        Object {
          "action": "control_end_thread",
          "values": Array [],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`функции обычные 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "call_function",
          "values": Array [
            Object {
              "name": "function_name",
              "value": Object {
                "text": "func",
                "type": "text",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
    Object {
      "is_hidden": false,
      "name": "func",
      "operations": Array [],
      "position": 1,
      "type": "function",
    },
  ],
}
`;

exports[`функции поздние 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "call_function",
          "values": Array [
            Object {
              "name": "function_name",
              "value": Object {
                "text": "func",
                "type": "text",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
    Object {
      "is_hidden": false,
      "name": "func",
      "operations": Array [],
      "position": 1,
      "type": "function",
    },
  ],
}
`;

exports[`функции процессы 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "start_process",
          "values": Array [
            Object {
              "name": "process_name",
              "value": Object {
                "text": "func",
                "type": "text",
              },
            },
            Object {
              "name": "local_variables_mode",
              "value": Object {
                "enum": "COPY",
                "type": "enum",
              },
            },
            Object {
              "name": "target_mode",
              "value": Object {
                "enum": "CURRENT_TARGET",
                "type": "enum",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
    Object {
      "is_hidden": false,
      "name": "func",
      "operations": Array [],
      "position": 1,
      "type": "process",
    },
  ],
}
`;

exports[`функции с аргументами 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "call_function",
          "values": Array [
            Object {
              "name": "function_name",
              "value": Object {
                "text": "func",
                "type": "text",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
    Object {
      "is_hidden": false,
      "name": "func",
      "operations": Array [],
      "position": 1,
      "type": "function",
    },
  ],
}
`;

exports[`функции с именными аргументами 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "call_function",
          "values": Array [
            Object {
              "name": "function_name",
              "value": Object {
                "text": "func",
                "type": "text",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
    Object {
      "is_hidden": false,
      "name": "func",
      "operations": Array [],
      "position": 1,
      "type": "function",
    },
  ],
}
`;

exports[`функции с опциональными аргументами 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 0,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "call_function",
          "values": Array [
            Object {
              "name": "function_name",
              "value": Object {
                "text": "func",
                "type": "text",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
    Object {
      "is_hidden": false,
      "name": "func",
      "operations": Array [],
      "position": 1,
      "type": "function",
    },
  ],
}
`;
